File: backend/models/Category.js
const mongoose = require('mongoose');

const categorySchema = new mongoose.Schema({
  name: {
    type: String,
    required: [true, 'Назва категорії є обов\'язковою'],
    unique: true,
    trim: true,
    maxlength: [50, 'Назва категорії не може перевищувати 50 символів']
  },
  description: {
    type: String,
    trim: true,
    maxlength: [200, 'Опис категорії не може перевищувати 200 символів']
  },
  color: {
    type: String,
    default: '#6B7280',
    match: [/^#[0-9A-F]{6}$/i, 'Колір повинен бути в форматі HEX']
  },
  icon: {
    type: String,
    trim: true,
    maxlength: [50, 'Назва іконки не може перевищувати 50 символів']
  },
  isActive: {
    type: Boolean,
    default: true
  },
  sortOrder: {
    type: Number,
    default: 0
  },
  createdBy: {
    type: mongoose.Schema.Types.ObjectId,
    ref: 'User',
    required: true
  },
  createdAt: {
    type: Date,
    default: Date.now
  },
  updatedAt: {
    type: Date,
    default: Date.now
  }
});

// Індекси
categorySchema.index({ name: 1 });
categorySchema.index({ isActive: 1, sortOrder: 1 });

// Middleware для оновлення updatedAt
categorySchema.pre('save', function(next) {
  this.updatedAt = new Date();
  next();
});

// Статичні методи
categorySchema.statics.getActiveCategories = function() {
  return this.find({ isActive: true }).sort({ sortOrder: 1, name: 1 });
};

categorySchema.statics.findByName = function(name) {
  return this.findOne({ name: new RegExp(`^${name}$`, 'i') });
};

// Методи екземпляра
categorySchema.methods.deactivate = function() {
  this.isActive = false;
  return this.save();
};

categorySchema.methods.activate = function() {
  this.isActive = true;
  return this.save();
};

module.exports = mongoose.model('Category', categorySchema);

File: mobile_flutter/lib/src/models/category.dart
class Category {
  final String id;
  final String name;
  final String? description;
  final String color;
  final String? icon;
  final int? sortOrder;
  final bool isActive;

  Category({
    required this.id,
    required this.name,
    this.description,
    required this.color,
    this.icon,
    this.sortOrder,
    required this.isActive,
  });

  factory Category.fromJson(Map<String, dynamic> json) {
    return Category(
      id: (json['_id'] ?? json['id'] ?? '').toString(),
      name: (json['name'] ?? '').toString(),
      description: json['description']?.toString(),
      color: (json['color'] ?? '#4F46E5').toString(),
      icon: json['icon']?.toString(),
      sortOrder: json['sortOrder'] is int ? json['sortOrder'] as int : int.tryParse('${json['sortOrder']}'),
      isActive: json['isActive'] == true || json['isActive'] == 'true',
    );
  }
}